### ISSUE TITLE
Fix simulation of 'click' event in drag action

### ISSUE THREAD

**Issue by helen-dikareva** [DevExpress/testcafe#2640]
We should check that we simulate/skip simulation for `click` event in the end of `drag` action just like the browser does. 
Take into account the following points:
1) prevention of previous events (mouseup, pointerup, touchup, etc)
2) if drag action is executed for really draggable elements or not
3) touch emulation specificity
4) different browsers specificity

**Comment by 0815Strohhut** [comment]
Hello, 

I'm hijacking this issue as I have a problem with the simulation of the click event. 

The situation: 
We are using a drag and drop solution (react-dnd-touch-backend) that does not use the native html5 api but instead has a custom handling for the drag and drop logic (supporting touch and click events). In our use case, the "preventDefault" is called on the "touchend" event, hence the browser (firefox/chrome) does not fire the click event (see  [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Touch_events/Supporting_both_TouchEvent_and_MouseEvent)). 
But, the testcafe event simulation does fire the click event anyway. 

Looking at src\client\automation\playback\drag\base.js -> _mouseup the issue seems to be that for

`eventSimulator[this.upEvent](topElement, options)`

the return value is ignored. What should probably happen is that if the return value is true, the click event at the end of the promise
`if (topElement && element === topElement && !this.dragAndDropState.enabled)
                            eventSimulator.click(topElement, options);
`
should be suppressed. 

I'd appreciate feedback on it, I can also try to provide a pull request.

**Comment by miherlosev** [comment]
Hi @0815Strohhut 

It's a very complex task. You need to emulate browser behavior for [all officially supported browsers](https://devexpress.github.io/testcafe/documentation/using-testcafe/common-concepts/browsers/browser-support.html#officially-supported-browsers). You can try to make a PR with the implementation of this functionality.

**Comment by 0815Strohhut** [comment]
Hi, 

I added a pull request for the first step: Handling touch events in a better way when touchend is cancelled.

That solution works for controlling the touch events when using a custom drag logic, but does not help with mouse events yet - as the click event is fired even when the mousedown/mouseup is cancelled. Hence, I am currently investigating the behavior for click events and my findings so far: 

Assume that you have a div containing two buttons (I created a [fiddle](https://jsfiddle.net/0815Strohhut/xsdLma67/52/show) to play around with it).

**Case 1**: Mousedown on button A, Mouseup on button A as well: 
-> all browsers seem to raise a click event on button A
**Case 2**: Mousedown on button A, Mouseup on button B: 
Depending on the browser, we see two different behaviors: 
1.  Raise the click event on the div containing the buttons (Chrome, Safari, Chromium Edge) 
2.  Do not raise a click event at all (legacy Edge, Firefox, IE) 

According to the [w3 spec](https://www.w3.org/TR/uievents/#events-mouseevent-event-order): 

> Each implementation will determine the appropriate hysteresis tolerance, but in general SHOULD fire click and dblclick events when the event target of the associated mousedown and mouseup events is the same element with no mouseout or mouseleave events intervening, and SHOULD fire click and dblclick events on the nearest common inclusive ancestor when the associated mousedown and mouseup event targets are different.

There are some old and open bugs for [firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1229143) and [chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=543776) on this where the result is more or less "we keep it as is". 

I haven't fully looked into the code yet, but assuming that we can determine the browser in the drag simulation, my idea would be:
1. For browsers in case 2.2, simulate click if and only if mousedown target = mouseup target
2. For browsers in case 2.1, find the first common ancestor (not sure what the spec means with "inclusive"?) and simulate the event on that ancestor
